### PR TITLE
Document hash verification

### LINKED ISSUE #24284: [Doc Handling] Add a cryptographic hash of the file contents to the document reference
**Description**

The first iteration of Document Handling does not cover authorization. Although process-definition-based authorizaiton will be handled centrally for the whole Camunda, we can introduce an intermediary solution to prevent attackers from accessing documents from a foreign process by brute-forcing the document ID.

- The document reference should include a cryptographic hash of the document content (e.g. SHA-256)
- In document content get requests, the hash value should be passed along with the document ID
- The server has to check the hash value to ensure only those who have access to the original document reference object can download the file.

### PR DESCRIPTION
**Pull request by mo-okocha**
## Checklist

- [ ] for CI changes:
  - [ ] structural/foundational changes signed off by [CI DRI](https://github.com/cmur2)
  - [ ] [ci.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml) modifications comply with ["Unified CI" requirements](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria)

## Related issues

closes #24284

### CODE REVIEW

File: zeebe/gateway-protocol/src/main/proto/rest-api.yaml
Comment (chillleader) [inline zeebe/gateway-protocol/src/main/proto/rest-api.yaml:1]: Here, we also need to extend the `DocumentReference` schema to include the hash property

File: clients/java/src/main/java/io/camunda/client/impl/CamundaClientImpl.java [hunk 1]
```diff
@@ -796,14 +797,21 @@ public CreateDocumentCommandStep1 newCreateDocumentCommand() {
...
-    return new DocumentContentGetRequestImpl(httpClient, documentId, null, config);
+    return new DocumentContentGetRequestImpl(httpClient, documentId, null, null, config);
   }
 
   @Override
   public DocumentContentGetRequest newDocumentContentGetRequest(
       final DocumentReferenceResponse documentReference) {
+    final String contentHash =
+        (String)
+            documentReference.getMetadata().getCustomProperties().get(CONTENT_HASH_METADATA_KEY);
```
Comment (chillleader) [inline clients/java/src/main/java/io/camunda/client/impl/CamundaClientImpl.java:808]: I think we should include the document hash into the root-level properties of the document reference response along with the document ID and store ID, not in custom properties of the metadata.

File: clients/java/src/main/java/io/camunda/client/impl/CamundaClientImpl.java [hunk 2]
```diff
@@ -796,14 +797,21 @@ public CreateDocumentCommandStep1 newCreateDocumentCommand() {
 
   @Override
   public DocumentContentGetRequest newDocumentContentGetRequest(final String documentId) {
-    return new DocumentContentGetRequestImpl(httpClient, documentId, null, config);
+    return new DocumentContentGetRequestImpl(httpClient, documentId, null, null, config);
```
Comment (chillleader) [inline clients/java/src/main/java/io/camunda/client/impl/CamundaClientImpl.java:800]: `DocumentContentGetRequest` should be extended to allow specifying the content hash. If the user creates the command using this method (only with document ID), it should be possible to specify content hash later in the builder.

File: clients/java/src/main/java/io/camunda/client/impl/fetch/DocumentContentGetRequestImpl.java [hunk 3]
```diff
@@ -63,6 +66,9 @@ public CamundaFuture<InputStream> send() {
     if (storeId != null) {
       queryParams.put("storeId", storeId);
     }
+    if (contentHash != null) {
+      queryParams.put("contentHash", contentHash);
+    }
```
Comment (chillleader) [inline clients/java/src/main/java/io/camunda/client/impl/fetch/DocumentContentGetRequestImpl.java:71]: Should we make this parameter required if the verification is enabled by default in all document stores?

**Review by chillleader — DISMISSED** [review]
Great job so far, thank you!

I think we need to adjust the solution to make hash a root-level property of the document reference and include it into the data schema, instead of passing it as a custom metadata entry.

Besides that, it would be great to add tests to cover this new funcitonality (at least some e2e test cases in the client implementation).

File: dist/src/main/java/io/camunda/application/StandaloneCamunda.java [hunk 4]
```diff
@@ -28,8 +25,7 @@
 public class StandaloneCamunda {
 
   private static final String SPRING_PROFILES_ACTIVE_PROPERTY = ACTIVE_PROFILES_PROPERTY_NAME;
-  private static final String DEFAULT_CAMUNDA_PROFILES =
```
Comment (marcosgvieira) [inline dist/src/main/java/io/camunda/application/StandaloneCamunda.java:31]: @mo-okocha let's not change StandaloneCamunda, I believe this changes were done for development tests, let's revert it.

File: dist/src/main/java/io/camunda/application/StandaloneGateway.java [hunk 5]
```diff
@@ -22,7 +23,10 @@ public static void main(final String[] args) {
 
     final var standaloneGatewayApplication =
         MainSupport.createDefaultApplicationBuilder()
-            .sources(CommonsModuleConfiguration.class, GatewayModuleConfiguration.class)
+            .sources(
+                CommonsModuleConfiguration.class,
+                GatewayModuleConfiguration.class,
+                BrokerModuleConfiguration.class)
```
Comment (marcosgvieira) [inline dist/src/main/java/io/camunda/application/StandaloneGateway.java:29]: @mo-okocha do we need to change here, why, can you check?

File: document/store/src/main/java/io/camunda/document/store/aws/AwsDocumentStore.java [hunk 6]
```diff
@@ -193,6 +209,33 @@ private Either<DocumentError, DocumentLink> linkDocumentInternal(
...
+      final HeadObjectResponse documentInfo = getDocumentInfo(documentId);
+      if (documentInfo == null) {
+        return Either.left(new DocumentNotFound(documentId));
+      }
+      if (!documentInfo.hasMetadata()) {
+        return Either.left(new DocumentError.InvalidInput("No metadata found for document"));
+      }
+
+      if (!documentInfo.metadata().containsKey(CONTENT_HASH_METADATA_KEY)) {
+        return Either.left(new DocumentError.InvalidInput("No content hash found for document"));
```
Comment (marcosgvieira) [inline document/store/src/main/java/io/camunda/document/store/aws/AwsDocumentStore.java:224]: @mo-okocha do we have unit tests for this?

**Review by marcosgvieira — CHANGES_REQUESTED** [review]
@mo-okocha good work, there are some changes in Gateway and CamundaStandalone that I think are related to development, please have a look.

Also there are missing unit tests, we should focus on implementing them.

File: document/store/src/main/java/io/camunda/document/store/aws/AwsDocumentStore.java (hunk 6 above)
Comment (mo-okocha) [reply document/store/src/main/java/io/camunda/document/store/aws/AwsDocumentStore.java:224]: Unit tests have been added for each branch in the method

**Review by marcosgvieira — APPROVED** [review]
lgtm!

**Pull request merged**